### DISCUSSION TITLE
Emgucv yolo vs darknet yolo

### DISCUSSION THREAD

**Discussion by Scalee** [emgucv/emgucv#415] [Q&A]
Looking at the table at darknet, opencv is supposed to beat darknet.
Darknet 100FPS and OpenCV 180FPS.

Using the YOLO example i am achieving about 100FPS, while darknet is running at 180FPS. 

Is this to be expected? Is this due to C# overhead?
Or am i just doing something wrong?

**Comment by emgucv** [comment]
Any refernce to that table? Can you post a link? 

What settings do they use? float32? Quantized? CUDA? Yolo version? Darknet version?

C# overhead is minimal when calling Open CV DNN module. It should perform very similar to native Open CV. Any code in C# that perform image rendering, such as converting Mat to Bitmap can cause additional overhead. But if both Darknet and Open CV use the same code to convert Mat to Bitmap and render, performance hit should affect both equally.

**Reply by Scalee** [reply]
The table can be found here, https://github.com/AlexeyAB/darknet#geforce-rtx-2080-ti 
As far as i understand it uses it's own inference engine, running on cuda and only uses opencv for image augmentation and preview.

I am using the same yolo4 config and weights file, using the latest version of darknet using the detect_mat method.
Running on a RTX 3080.

As for Mat converting, Emgucv should have a leg up here.
With Emgucv i only create a mat once with a pointer to the pixel data and overwrite the pixel data at the pointer every frame and pass that to ``net.SetInput(DnnInvoke.BlobFromImage(img, 1 / 255.0, new Size(size.X, size.Y), new Emgu.CV.Structure.MCvScalar(), true, false), "data");``

While darknet creates a new one and uses imdecode:
Snippet from detect_mat function
``
std::vector<char> vdata(data, data + data_length);
cv::Mat image = imdecode(cv::Mat(vdata), 1);

std::vector<bbox_t> detection = detector->detect(image);
``

[https://i.ibb.co/6sxP70W/Screenshot-2021-02-08-185745.png](url)
These 3 functions seems to up most of the time, i would need to get them allot faster if i want to match the 180fps.

**Reply by PPS100** [reply]
I am also working on darknet/yolo and engu......Looking for someone to bounce ideas with and help me with some challenges I am having. Will to pay 50/hr USD for some help.  My twitter is @borgenp
